### PR TITLE
refactor: [M3-7920, M3-7930] – Refactor VPC queries

### PR DESCRIPTION
**Pull request by dwiley-akamai**
## Description 📝
Refactor the VPC queries to use `createQueryKeys` from `query-key-factory`.

The main objective was to achieve parity with existing functionality, although some small enhancements at the margins were done as well.

## Changes  🔄
- Use `createQueryKeys` to generate VPC & subnet query keys
- Update mock region data to include VPC capability
- Add new `useAllVPCsQuery` and use it in `VPCPanel.tsx` & `useVPCConfigInterface.tsx`
- Remove manual uses of `vpcQueryKey` and `subnetQueryKey` throughout codebase, notably for various query invalidations
- Remove unused `useSubnetQuery`

## Target release date 🗓️
4/15/24

### Verification steps
Essentially the entire VPC offering in Cloud needs to be tested to ensure nothing has broken with these changes.

- VPC landing page
  - query for paginated results should refresh upon VPC creation
- Assign/Unassign Linode flow via drawer accessed from Subnets table
  - should invalidate the paginated VPC, individual VPC, and VPC subnets queries
- Assigning Linode to VPC via Linode Config dialog
  - all VPC queries should be invalidated
- Linode Create flow
  - a newly-created linode assigned to a VPC should have the VPC section present in the Linode Detail header
- Linode Detail page
  - the VPC section in the Linode Detail header should be accurate, and the "IP Addresses" table on the "Network" tab should contain the expected VPC-related IPs
- Deleting a Linode
  - the 'all' VPCs, paginated VPCs, individual VPC, and VPC subnets queries should be invalidated upon Linode deletion

## As an Author I have considered 🤔
- [X] 👀 Doing a self review
- [X] ❔ Our [contribution guidelines](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md)
- [X] 🤏 Splitting feature into small PRs
- [X] ➕ Adding a changeset
- [X] 🧪 Providing/Improving test coverage
- [X] 🔐 Removing all sensitive information from the code and PR description
- [X] 🚩 Using a feature flag to protect the release
- [X] 👣 Providing comprehensive reproduction steps
- [X] 📑 Providing or updating our documentation
- [X] 🕛 Scheduling a pair reviewing session
- [X] 📱 Providing mobile support
- [X] ♿  Providing accessibility support

### CODE REVIEW

File: packages/manager/src/queries/vpcs/vpcs.ts [hunk 1]
```diff
@@ -16,68 +16,110 @@ import {
...
+      }),
+      subnets: {
+        paginated: (params: Params = {}, filter: Filter = {}) => ({
+          queryFn: () => getSubnets(vpcId, params, filter),
+          queryKey: [params, filter],
+        }),
+      },
+    },
+    queryFn: () => getVPC(vpcId),
+    queryKey: [vpcId],
```
Comment (dwiley-akamai) [inline packages/manager/src/queries/vpcs/vpcs.ts:50]: @bnussman-akamai Does this look correct to you? For some reason, when I go into `useSubnetsQuery` and begin refactoring it, I type `...vpcQueries.vpc(vpcID).` but `_ctx` doesn't get picked up on as something that exists on the object. 

I pulled an example from the docs,

```
export const users = createQueryKeys('users', {
  detail: (userId: string) => ({
    contextQueries: {
      likes: {
        queryFn: () => api.getUserLikes(userId),
        queryKey: null,
      },
    },
    queryFn: () => api.getUser(userId),
    queryKey: [userId],
  }),
});
```

just to test the autocomplete and see if there's a syntax issue, but it successfully picked up on `users.detail(userId)._ctx` as expected.

Am I overlooking something here?

Comment (bnussman-akamai) [reply packages/manager/src/queries/vpcs/vpcs.ts:50]: Give this a try

```typescript
export const vpcQueries = createQueryKeys('vpcs', {
  vpc: (vpcId: number) => ({
    contextQueries: {
      subnets: {
        contextQueries: {
          paginated: (params: Params = {}, filter: Filter = {}) => ({
            queryFn: () => getSubnets(vpcId, params, filter),
            queryKey: [params, filter],
          }),
          subnet: (subnetId: number) => ({
            queryFn: () => getSubnet(vpcId, subnetId),
            queryKey: [subnetId],
          }),
        },
        queryKey: null,
      },
    },
    queryFn: () => getVPC(vpcId),
    queryKey: [vpcId],
  }),
  vpcs: {
    contextQueries: {
      all: {
        queryFn: getAllVPCsRequest,
        queryKey: null,
      },
      paginated: (params: Params = {}, filter: Filter = {}) => ({
        queryFn: () => getVPCs(params, filter),
        queryKey: [params, filter],
      }),
    },
    queryKey: null,
  },
});
```

- I added `queryKey: [subnetId]` to the subnet definition
- I added more nesting to the subnets queries

File: packages/manager/src/queries/vpcs/vpcs.ts
Comment (dwiley-akamai) [inline packages/manager/src/queries/vpcs/vpcs.ts:1]: This query previously existed:

```
export const useSubnetQuery = (vpcID: number, subnetID: number) => {	
  return useQuery<Subnet, APIError[]>(	
    [vpcQueryKey, 'vpc', vpcID, subnetQueryKey, 'subnet', subnetID],	
    () => getSubnet(vpcID, subnetID)	
  );	
};
```

Most of the original queries were written before actual dev work on the UI began last summer, and it turns out this was never used (there's nothing approximating a Subnet Details page) so I removed it.

File: packages/manager/src/hooks/useUnassignLinode.ts [hunk 2]
```diff
@@ -1,20 +1,19 @@
...
+import { vpcQueries } from 'src/queries/vpcs/vpcs';
 
 import type {
   APIError,
   DeleteLinodeConfigInterfacePayload,
 } from '@linode/api-v4';
 
-type IdsForUnassignLinode = DeleteLinodeConfigInterfacePayload & {
-  subnetId: number;
+interface IdsForUnassignLinode extends DeleteLinodeConfigInterfacePayload {
```
Comment (dwiley-akamai) [inline packages/manager/src/hooks/useUnassignLinode.ts:14]: match the new convention (see https://github.com/linode/manager/pull/10309 for more)

File: packages/manager/src/hooks/useUnassignLinode.ts [hunk 3]
```diff
@@ -1,20 +1,19 @@
...
-import { subnetQueryKey, vpcQueryKey } from 'src/queries/vpcs';
+import { vpcQueries } from 'src/queries/vpcs/vpcs';
 
 import type {
   APIError,
   DeleteLinodeConfigInterfacePayload,
 } from '@linode/api-v4';
 
-type IdsForUnassignLinode = DeleteLinodeConfigInterfacePayload & {
-  subnetId: number;
```
Comment (dwiley-akamai) [inline packages/manager/src/hooks/useUnassignLinode.ts:15]: Nothing at the level of `[vpcQueryKey, 'vpc', vpcId, subnetQueryKey, 'subnet', subnetId]` was ever used in the actual code so this was superfluous

File: packages/manager/src/queries/vpcs/vpcs.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+            queryKey: [subnetId],
+          }),
+        },
+        queryKey: null,
+      },
+    },
+    queryFn: () => getVPC(vpcId),
+    queryKey: [vpcId],
+  }),
+  vpcs: {
```
Comment (dwiley-akamai) [inline packages/manager/src/queries/vpcs/vpcs.ts:55]: This results in `"vpcs","vpcs" [...]` which I do not like given the redundancy, but I think we have to accept?

Screenshot 2024-04-02 at 3 52 16 PM

Comment (bnussman-akamai) [reply packages/manager/src/queries/vpcs/vpcs.ts:52]: Could we "un-nest" the `all` and `paginated` queries?

Like this:

```typescript
export const vpcQueries = createQueryKeys('vpcs', {
  all: {
    queryFn: getAllVPCsRequest,
    queryKey: null,
  },
  paginated: (params: Params = {}, filter: Filter = {}) => ({
    queryFn: () => getVPCs(params, filter),
    queryKey: [params, filter],
  }),
  vpc: (vpcId: number) => ({
    contextQueries: {
      subnets: {
        contextQueries: {
          paginated: (params: Params = {}, filter: Filter = {}) => ({
            queryFn: () => getSubnets(vpcId, params, filter),
            queryKey: [params, filter],
          }),
          subnet: (subnetId: number) => ({
            queryFn: () => getSubnet(vpcId, subnetId),
            queryKey: [subnetId],
          }),
        },
        queryKey: null,
      },
    },
    queryFn: () => getVPC(vpcId),
    queryKey: [vpcId],
  }),
})
```

Comment (dwiley-akamai) [reply packages/manager/src/queries/vpcs/vpcs.ts:52]: @bnussman-akamai Yes, that works and allows us to clean things up a bit 👍🏾 Going to push that change

File: packages/manager/src/queries/vpcs/vpcs.ts [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+export const useDeleteSubnetMutation = (vpcId: number, subnetId: number) => {
+  const queryClient = useQueryClient();
+  return useMutation<{}, APIError[]>({
+    mutationFn: () => deleteSubnet(vpcId, subnetId),
+    onSuccess: () => {
+      // New subnet created --> refresh the VPC queries (all, paginated, & individual), plus the /subnets VPC query
+      queryClient.invalidateQueries(vpcQueries.all);
+      queryClient.invalidateQueries(vpcQueries.paginated._def);
+      queryClient.invalidateQueries(vpcQueries.vpc(vpcId).queryKey);
+      queryClient.invalidateQueries(vpcQueries.vpc(vpcId)._ctx.subnets);
```
Comment (bnussman-akamai) [inline packages/manager/src/queries/vpcs/vpcs.ts:166]: Unfortunately typescript doesn't catch these causes, but can we update this to be more explicit? Add `queryKey` to the end. This change might be needed in a few places.

```suggestion
      queryClient.invalidateQueries(
        vpcQueries.vpc(vpcId)._ctx.subnets.queryKey
      );
```

File: packages/manager/src/hooks/useVPCConfigInterface.ts
[the remark this answers is not in the document]
Comment (dwiley-akamai) [reply packages/manager/src/hooks/useVPCConfigInterface.ts:1]: We previously used `useVPCsQuery` here, but I think that would pose an issue for some data on the Linode Detail page (see `LinodeEntityDetail.tsx` and `LinodeIPAddress.tsx`) for customers with lots of VPCs if the VPC wasn't on the first page of results?

File: packages/manager/src/features/Linodes/LinodesDetail/LinodeConfigs/LinodeConfigDialog.tsx [hunk 6]
```diff
@@ -402,7 +402,7 @@ export const LinodeConfigDialog = (props: Props) => {
           (thisInterface) => thisInterface.purpose === 'vpc'
         )
       ) {
-        queryClient.invalidateQueries([vpcQueryKey]);
+        queryClient.invalidateQueries(vpcQueries._def);
```
[the remark this answers is not in the document]
Comment (dwiley-akamai) [reply packages/manager/src/features/Linodes/LinodesDetail/LinodeConfigs/LinodeConfigDialog.tsx:405]: That's fair, I think that level of granularity catches everything we'd need invalidated.

File: packages/manager/src/hooks/useVPCConfigInterface.ts
[the remark this answers is not in the document]
Comment (dwiley-akamai) [reply packages/manager/src/hooks/useVPCConfigInterface.ts:1]: Whoops, I misread your original comment. Yes, that worked fine from my testing so I pushed the refactor in the most recent commit.

**Comment by cliu-akamai** [comment]
> @cliu-akamai There's an E2E failure for `linode-config.spec.ts` (failed on the previous run and fails locally) that I haven't been able to diagnose -- could you take a look at that?

Sure thing!

**Comment by cliu-akamai** [comment]
If we stop querying the VPCs when accessing endpoint `/linodes/${linodeId}/configurations`, we should remove Lines [149](https://github.com/dwiley-akamai/manager/blob/M3-7920-7930-vpc-queries-updates/packages/manager/cypress/e2e/core/linodes/linode-config.spec.ts#L149) and [304](https://github.com/dwiley-akamai/manager/blob/M3-7920-7930-vpc-queries-updates/packages/manager/cypress/e2e/core/linodes/linode-config.spec.ts#L304) in `linode-config.spec.ts` to avoid test failure.

**Comment by bnussman-akamai** [comment]
Looks like deleting lines `138`, `149`, `292`, and `303` fix the failures. I think this makes sense because of the optimization we made in `useVPCConfigInterface`.  We no longer need to `GET /v4/vpcs` because we now just fetch the singular VPC using  `GET /v4/vpcs/:id`

I'm happy to push that change if you want me to!

**Comment by dwiley-akamai** [comment]
> Looks like deleting lines `138`, `149`, `292`, and `303` fix the failures. I think this makes sense because of the optimization we made in `useVPCConfigInterface`. We no longer need to `GET /v4/vpcs` because we now just fetch the singular VPC using `GET /v4/vpcs/:id`

Initially it was a bit confusing because `queryClient.invalidateQueries(vpcQueries.paginated._def)` in `LinodeConfigDialog.tsx` seemed to correspond with the `@getVPCs` in the test, but in terms of the actual sequence of interactions it makes sense (that invalidation doesn't happen until the user clicks "Save Changes" and it goes through successfully). So I think we're good on that aspect 👍🏾 cc @cliu-akamai

Another point is looking into removing the `mockAppendFeatureFlags` logic in `linode-config.spec.ts` since VPC isn't feature flagged anymore. That's not in-scope for this ticket, but commenting those pieces out does make the test fail again locally

File: packages/manager/src/hooks/useUnassignLinode.ts [hunk 7]
```diff
@@ -30,14 +29,12 @@ export const useUnassignLinode = () => {
...
     linodeId,
-    subnetId,
     vpcId,
   }: InvalidateSubnetLinodeConfigQueryIds) => {
     const queryKeys = [
-      [vpcQueryKey, 'paginated'],
-      [vpcQueryKey, 'vpc', vpcId],
-      [vpcQueryKey, 'vpc', vpcId, subnetQueryKey],
-      [vpcQueryKey, 'vpc', vpcId, subnetQueryKey, 'subnet', subnetId],
+      vpcQueries.paginated._def,
```
Comment (bnussman-akamai) [inline packages/manager/src/hooks/useUnassignLinode.ts:35]: Should `vpcQueries.all.queryKey` be invalidated here?

File: packages/manager/src/features/Linodes/LinodesCreate/LinodeCreateContainer.tsx [hunk 8]
```diff
@@ -911,7 +911,7 @@ class LinodeCreateContainer extends React.PureComponent<CombinedProps, State> {
           this.state.selectedVPCId !== undefined &&
           this.state.selectedVPCId !== -1
         ) {
-          this.props.queryClient.invalidateQueries([vpcQueryKey, 'paginated']);
+          this.props.queryClient.invalidateQueries(vpcQueries.all.queryKey);
```
Comment (bnussman-akamai) [inline packages/manager/src/features/Linodes/LinodesCreate/LinodeCreateContainer.tsx:914]: I'm wondering if we should also invalidate `vpcQueries.paginated._def` and `vpcQueries.vpc(selectedVPCId).queryKey` here

**Review by bnussman-akamai — APPROVED** [review]
Looks great 🔥  Thanks for doing this!

**Review by cpathipa — APPROVED** [review]
Nice work @dwiley-akamai! 
confirming on the VPC flow and no regressions found.

**Pull request merged**